### PR TITLE
sharded_jit cleanup

### PR DESCRIPTION
**Pull request by skye**
* Add sharding utilities to xla_bridge
* Change partitions argument to in_parts and out_parts
* Add unit tests
* Reuse more pxla functionality
* Remove stale translation rule
* Fail on non-TPU platforms
* Add docstring
* And more!

### CODE REVIEW

**Comment by skye** [comment]
@jekbradbury sorry this is so long, lemme know if I should break it up into smaller changes.

I verified that the tests pass internally. I think we should be able to run these on Cloud TPUs too, but it requires a jaxlib change to enable SPMD partitioning without the flag.

File: jax/lib/xla_bridge.py [hunk 1]
```diff
@@ -292,6 +292,41 @@ def constant(builder, py_val, canonicalize_types=True):
   else:
     raise TypeError("No constant handler for type: {}".format(py_type))
 
+# XLA supports arbitrary tuple nesting, but JAX only uses one level of tupling
+# (and our type checkers don't support recursive types).
+SpatialSharding = Union[Sequence[int], None, Tuple[Sequence[int], None]]
+
+def _sharding_to_proto(sharding: SpatialSharding):
```
Comment (gnecula) [inline jax/lib/xla_bridge.py:299]: I cannot tell what this code is supposed to do, since one would need to understand the OpSharding proto. I wonder if it is easy to explain the requirements on the input, e.g., if the input is a sequence of integers, it must have the same length as the data rank, and the product is the number of devices?

File: jax/lib/xla_bridge.py [hunk 2]
```diff
@@ -292,6 +292,41 @@ def constant(builder, py_val, canonicalize_types=True):
...
+  if sharding is None:
+    proto.type = xla_client.OpSharding.Type.REPLICATED
+  else:
+    proto.type = xla_client.OpSharding.Type.OTHER
+    proto.tile_assignment_dimensions = list(sharding)
+    proto.tile_assignment_devices = list(range(onp.product(sharding)))
+  return proto
+
+def set_sharding(builder, op, sharding: SpatialSharding):
+  """Uses CustomCall to annotate a value as sharded."""
```
Comment (gnecula) [inline jax/lib/xla_bridge.py:318]: Is this an existing customcall? What does it do?

File: tests/sharded_jit_test.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+from jax.interpreters.sharded_jit import PartitionSpec as P
+
+from jax.config import config
+config.parse_flags_with_absl()
+
+
+class ShardedJitTest(jtu.JaxTestCase):
+
+  def testBasic(self):
+    if jtu.device_under_test() != "tpu" or jax.device_count() < 2:
```
Comment (gnecula) [inline tests/sharded_jit_test.py:39]: Maybe move this into `def setUp(self)` if it is common for all tests?

File: tests/sharded_jit_test.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      a1, a2 = a
+      c1, (c2, c3) = c
+      return a1 + a2 + b + c1 + c2 + c3
+
+    def _make_arg(*shape):
+      return onp.arange(onp.prod(shape)).reshape(shape)
+
+    a = (_make_arg(4, 4), _make_arg())
+    b = _make_arg(4, 4)
+    c = [_make_arg(), (_make_arg(4, 4), _make_arg(4, 4))]
```
Comment (gnecula) [inline tests/sharded_jit_test.py:70]: Is `make_arg` producing a 0?

File: tests/sharded_jit_test.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    in_parts = (None, P(2, 1), [None, P(2, 1)])
```
Comment (gnecula) [inline tests/sharded_jit_test.py:72]: I do not quite understand how to make the `in_parts` for `c` to the structure of `c`, specifically the second element of a list: it is a pair, but it gets only a single sharding specification `P(2, 1)`. Wouldn't it be simpler to request
that the `in_parts` matches the structure of the argument?

File: tests/sharded_jit_test.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+      raise SkipTest
+
+    @partial(sharded_jit, in_parts=None, out_parts=None)
+    def f(x):
+      return x + 1
+
+    result = f(1.)
+    self.assertEqual(result, 2.)
+    self.assertIsInstance(result, pxla.ShardedDeviceArray)
+    self.assertLen(result.device_buffers, 1)
```
Comment (gnecula) [inline tests/sharded_jit_test.py:119]: I do not understand in what way is this replicated? (the name of the test is `testAllReplicated`)

File: jax/interpreters/sharded_jit.py [hunk 7]
```diff
@@ -279,19 +192,78 @@ def _sharded_call_impl(fun: lu.WrappedFun, *args, **params):
...
+  run across all devices and each device runs only a shard of the full
+  data. (Some data can optionally be replicated, which is sometimes more
+  efficient for small arrays when combined with larger spatially-partitioned
+  arrays.) Communication between devices is automatically inserted as necessary.
+
+  ``sharded_jit`` can be useful if the jitted version of ``fun`` would not fit
+  in a single device's memory, or to speed up ``fun`` by running each operation
+  in parallel across multiple devices.
+
+  Note: ``sharded_jit`` is currently available on TPU only!
```
[the remark this answers is not in the document]
Comment (gnecula) [reply jax/interpreters/sharded_jit.py:224]: I saw the check later.

File: jax/lib/xla_bridge.py (hunk 1 above)
Comment (skye) [reply jax/lib/xla_bridge.py:305]: I expanded the comment on the SpatialSharding type annotation since that's used in multiple functions here, and linked to the OpSharding proto here (which could probably be more clearly documented itself...). Let me know if specific aspects are still confusing.

File: jax/lib/xla_bridge.py (hunk 2 above)
Comment (skye) [reply jax/lib/xla_bridge.py:330]: It is. I added a comment.

File: tests/sharded_jit_test.py (hunk 3 above)
Comment (skye) [reply tests/sharded_jit_test.py]: Good idea, done.

File: tests/sharded_jit_test.py (hunk 4 above)
Comment (skye) [reply tests/sharded_jit_test.py:73]: Yes, `np.prod(())` returns 1.0 (so this is actually 0.).

File: tests/sharded_jit_test.py (hunk 5 above)
Comment (skye) [reply tests/sharded_jit_test.py:75]: I'm reusing the vmap machinery (`api_util.flatten_axes`, which I just realized I should rename now :)). I think it's useful to allow a pytree prefix, rather than the exact pytree structure, in the case where you have a complicated pytree (e.g. a flax model) and want to partition or replicate the whole thing.

Relatedly, I think it's gonna be hard if you do have something like a flax model, and want to partition just one part of it, or similar vmap just one part of it. A problem for another day though.

File: tests/sharded_jit_test.py (hunk 6 above)
Comment (skye) [reply tests/sharded_jit_test.py:119]: The test name is referring to all inputs/outputs being specified as replicated, but you're right that this isn't meaningfully replicated. I renamed the test. This is kind of a funny case, but I think running a single "replica" is the only reasonable behavior (or possibly requiring that a separate `num_partitions` argument is provided, which I'm gonna add later to support this case with partitioned infeed).

File: jax/interpreters/sharded_jit.py [hunk 8]
```diff
@@ -279,19 +188,84 @@ def _sharded_call_impl(fun: lu.WrappedFun, *args, **params):
...
+    fun: Function to be jitted.
+    in_parts: The input partitions, i.e. how each argument to ``fun`` should be
+      partitioned or replicated. This should be a PartitionSpec indicating into
+      how many partitions each dimension should be sharded, None indicating
+      replication, or (nested) standard Python containers thereof. For example,
+      ``in_parts=PartitionSpec(2,1)`` means all arguments should be partitioned
+      over two devices across the first dimension;
+      ``in_parts=(PartitionSpec(2,2), PartitionSpec(4,1), None)`` means the
+      first argument should be partitioned over four devices by splitting the
+      first two dimensions in half, the second argument should be partitioned
```
Comment (jekbradbury) [inline jax/interpreters/sharded_jit.py:239]: I think I'd support a third kind of `in_part`: an integer, meaning partition that numbered axis `n_partitions` ways (especially if that's the 90% use case).

File: tests/sharded_jit_test.py (hunk 5 above)
Comment (jekbradbury) [reply tests/sharded_jit_test.py:75]: Yeah, I think fundamentally that's the "API for transforming parameterized/stateful functions" problem and we'll need to solve it somewhere higher.

File: jax/interpreters/sharded_jit.py [hunk 9]
```diff
@@ -13,262 +13,172 @@
...
-  subc._builder.set_sharding(_sharding_to_proto(partitions[1]))
-  out_tuple = subc.Tuple(*out_nodes)
-  subc._builder.clear_sharding()
+  if partitions is None:
+    return pxla.ShardingSpec(
+        # int(1) because pytype is confused by 1 (???)
+        shards_per_axis=(int(1),) * len(aval.shape),
+        is_axis_materialized=(True,) * len(aval.shape),
+        replication_factor=num_partitions)
+  else:
```
Comment (jekbradbury) [inline jax/interpreters/sharded_jit.py:139]: Does the XLA partitioner support partially-sharded, partially-replicated arguments? I'm guessing no?

File: jax/lib/xla_bridge.py [hunk 10]
```diff
@@ -292,6 +292,55 @@ def constant(builder, py_val, canonicalize_types=True):
   else:
     raise TypeError("No constant handler for type: {}".format(py_type))
 
+# HLO instructions optionally can be annotated to say how the output should be
+# spatially partitioned (represented in XLA as OpSharding protos, see
+# _sharding_to_proto). For array outputs, the annotation is either an int per
+# dimension specifying the number of ways that dimension divided (meaning the
+# total number of shards in the product), or None to indicate the array should
+# be replicated. Tuple outputs are represented as tuples thereof. XLA supports
+# arbitrary tuple nesting, but JAX only uses one level of tupling (and our type
```
Comment (jekbradbury) [inline jax/lib/xla_bridge.py:301]: This wouldn't have to have the outer level of tupling if `tuple_args` shenanigans were pushed into PJRT, right?

File: jax/interpreters/sharded_jit.py (hunk 8 above)
Comment (jekbradbury) [reply jax/interpreters/sharded_jit.py:239]: (Although in that case I'd maybe want the kwargs to be `in_axes` and `out_axes` 🙂 )

**Review by jekbradbury — APPROVED** [review]
This is great! Merge whenever you're ready 🙂

File: jax/lib/xla_bridge.py (hunk 10 above)
Comment (skye) [reply jax/lib/xla_bridge.py:301]: I also use it for the outputs tuple, which I think we'll always need to build directly. (We also build the inputs tuple ourselves now, not sure if there are plans to have PJRT rewrite the actual HLO parameters into a tuple.) Although I could probably use a custom_call instead of putting the annotation directly on the inputs/outputs, and then I wouldn't need tuples. I'm gonna keep this since it's already done though.

File: jax/interpreters/sharded_jit.py (hunk 8 above)
Comment (skye) [reply jax/interpreters/sharded_jit.py:239]: Note there is no `n_partitions` argument yet :P I think this is a good idea though. One could also write a wrapper function (maybe? It gets confusing since you currently need to know the rank of the arguments, although we could relax that too).

File: jax/interpreters/sharded_jit.py (hunk 9 above)
Comment (skye) [reply jax/interpreters/sharded_jit.py:139]: You mean like e.g., cut in half then replicate twice? Pretty sure no. At least I'm not sure how you would express that with an OpSharding. I think regular replication (i.e. with pmap) + sharding achieves the same thing though?

File: jax/interpreters/sharded_jit.py [hunk 11]
```diff
@@ -13,262 +13,172 @@
...
+  if len(partition_specs) == 0:
+    # Everything is specified as replicated (all Nones).
+    return 1
+  num_partitions_set = set(np.prod(spec) for spec in partition_specs)
   if len(num_partitions_set) > 1:
     raise ValueError(
-        "All partition specs must use the same number of total partitions, "
-        "got: %s" % partitions)
+        f"All partition specs must use the same number of total partitions, "
+        "got {partitions}, with distinct number of partitions "
```
Comment (gnecula) [inline jax/interpreters/sharded_jit.py:171]: You have to use "f" before each string literal that contains interpolation in an implicit concatenation (I believe)

File: tests/sharded_jit_test.py (hunk 4 above)
Comment (gnecula) [reply tests/sharded_jit_test.py:73]: Then wouldn't it be easier to see what is going on by using 0 instead of `make_arg()`

File: tests/sharded_jit_test.py (hunk 5 above)
Comment (gnecula) [reply tests/sharded_jit_test.py:75]: Let's add some documentation somewhere, e.g., in the FAQ, or the pytree page, with a notes from the various API docstrings.

File: tests/sharded_jit_test.py (hunk 4 above)
Comment (skye) [reply tests/sharded_jit_test.py]: Yup, done! (I made them non-zero ints for good measure.)

File: tests/sharded_jit_test.py (hunk 5 above)
Comment (skye) [reply tests/sharded_jit_test.py:75]: I filed https://github.com/google/jax/issues/3095 to keep track of this. I'd prefer not to hold up this PR on adding + reviewing the new docs, plus it should probably be its own PR anyway. I can take that issue once I get the rest of my sharded_jit PRs are ready, but I left it unassigned in the meantime.

File: jax/interpreters/sharded_jit.py (hunk 11 above)
Comment (skye) [reply jax/interpreters/sharded_jit.py]: Done.

**Comment by skye** [comment]
@gnecula LGTM's offline. I also verified that the tests still pass on TPU.
